src/grid/storage.rs

```diff
@@ -34,86 +36,81 @@ impl<T: PartialEq> ::std::cmp::PartialEq for Storage<T> {
 
 impl<T> Storage<T> {
     #[inline]
-    pub fn with_capacity(cap: usize, lines: Line) -> Storage<T> {
+    pub fn with_capacity(cap: usize, lines: Line, template: T) -> Storage<T>
+    where
+        T: Clone,
+    {
+        // Allocate all lines in the buffer, including scrollback history
+        //
+        // TODO (jwilm) Allocating each line at this point is expensive and
+        // delays startup. A nice solution might be having `Row` delay
+        // allocation until it's actually used.
+        let inner = vec![template; cap];
         Storage {
-            inner: Vec::with_capacity(cap),
+            inner,
             zero: 0,
             visible_lines: lines - 1,
+            len: cap,
         }
     }
 
-    #[inline]
-    pub fn capacity(&self) -> usize {
-        self.inner.capacity()
-    }
-
     /// Increase the number of lines in the buffer
     pub fn grow_visible_lines(&mut self, next: Line, template_row: T)
```

Comment:
It's too bad we need to take a `T` here; it would be nice if we could avoid this allocation in the event that the buffer is already the right size.

Reply:
We could probably take `&T` here because `vec!` makes use of `Clone` anyways, but we would still need to allocate it at the call site anyways and in that case we would actually have one additional allocation in case the buffer is already the right size, so no gain there.